dll/win32/msv1_0/sam.c

```diff
@@ -454,3 +454,126 @@ SamValidateNormalUser(
+    _Out_ SAMPR_HANDLE* UserHandlePtr,
+    _Out_ PSAMPR_USER_INFO_BUFFER* UserInfoPtr,
+    _Out_ PNTSTATUS SubStatus)
+{
+    static const UNICODE_STRING NtAuthorityU = RTL_CONSTANT_STRING(L"NT AUTHORITY");
+    static const UNICODE_STRING LocalServiceU = RTL_CONSTANT_STRING(L"LocalService");
+    static const UNICODE_STRING NetworkServiceU = RTL_CONSTANT_STRING(L"NetworkService");
+
+    NTSTATUS Status = STATUS_SUCCESS;
+
+    *SpecialAccount = FALSE;
+
+    /* Check for special accounts */
+    // FIXME: Windows does not do this that way!! (msv1_0 does not contain these hardcoded values)
+    if (RtlEqualUnicodeString(LogonDomain, &NtAuthorityU, TRUE))
+    {
+        *SpecialAccount = TRUE;
+
+        /* Get the authority domain SID */
+        Status = GetNtAuthorityDomainSid(AccountDomainSidPtr);
+        if (!NT_SUCCESS(Status))
+        {
+            ERR("GetNtAuthorityDomainSid() failed (Status 0x%08lx)\n", Status);
+            return Status;
+        }
+
+        if (RtlEqualUnicodeString(LogonUserName, &LocalServiceU, TRUE))
+        {
+            TRACE("SpecialAccount: LocalService\n");
+
+            if (LogonType != Service)
+                return STATUS_LOGON_FAILURE;
+
+            *UserInfoPtr = RtlAllocateHeap(RtlGetProcessHeap(),
+                                       HEAP_ZERO_MEMORY,
+                                       sizeof(SAMPR_USER_ALL_INFORMATION));
+            if (*UserInfoPtr == NULL)
+            {
+                Status = STATUS_INSUFFICIENT_RESOURCES;
+                goto done;
+            }
+
+            (*UserInfoPtr)->All.UserId = SECURITY_LOCAL_SERVICE_RID;
+            (*UserInfoPtr)->All.PrimaryGroupId = SECURITY_LOCAL_SERVICE_RID;
+        }
+        else if (RtlEqualUnicodeString(LogonUserName, &NetworkServiceU, TRUE))
+        {
+            TRACE("SpecialAccount: NetworkService\n");
+
+            if (LogonType != Service)
+                return STATUS_LOGON_FAILURE;
+
+            *UserInfoPtr = RtlAllocateHeap(RtlGetProcessHeap(),
+                                       HEAP_ZERO_MEMORY,
+                                       sizeof(SAMPR_USER_ALL_INFORMATION));
+            if (*UserInfoPtr == NULL)
+            {
+                Status = STATUS_INSUFFICIENT_RESOURCES;
+                goto done;
+            }
+
+            (*UserInfoPtr)->All.UserId = SECURITY_NETWORK_SERVICE_RID;
+            (*UserInfoPtr)->All.PrimaryGroupId = SECURITY_NETWORK_SERVICE_RID;
+        }
+        else
+        {
+            Status = STATUS_NO_SUCH_USER;
+            goto done;
+        }
+    }
+    else
+    {
+        TRACE("NormalAccount\n");
+        Status = SamValidateNormalUser(
+            LogonUserName,
+            LogonPwdData,
+            ComputerName,
+            AccountDomainSidPtr,
+            UserHandlePtr,
+            UserInfoPtr,
+            SubStatus);
+        if (!NT_SUCCESS(Status))
+        {
+            ERR("SamValidateUser() failed (Status 0x%08lx)\n", Status);
```

Comment:
Should this say `SamValidateNormalUser() failed`?

Reply:
Fixed typo. Thanks.